README
SunUO README
============

(c) 2005-2007 Max Kellermann <max@duempel.org>

$Id$


What is SunUO?
--------------

SunUO is a free Ultima Online server, based on RunUO.


Getting SunUO
-------------

You can download SunUO on the home page:

 http://www.sunuo.org/


Installation and Configuration
------------------------------

Most of the documentation has moved to another file written in
DocBook. There is an online version:

 http://www.sunuo.org/doc/sunuo.html


For the impatient
-----------------

1. unpack the SunUO binary zip file

2. edit "etc/sunuo.xml" and enter the correct UO client data path

3. start "run.bat" (Windows/.NET) or "run.sh" (Linux/Mono)


Legal
-----

Copyright 2005-2007 Max Kellermann <max@duempel.org>
based on RunUO (C) The RunUO Software Team

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


NEWS
sunuo (0.5.2) unstable; urgency=low

  * when closing gumps, use the correct type id and remove gump from the
    internal gump list (patch from Defiance)
  * second fix for the PlayerVendor cash bug, perform checked addition to
    detect integer overflows (patch from Patrick Reinhardt)
  * don't grant refund for deleted customized houses (reported by pmb)
  * use .NET 2.0 for the windows build

 --

sunuo (0.5.1) unstable; urgency=low

  * impose restrictions on the invisibility spell (patch from Kai
    Sassmannshausen):
    - don't allow players to hide staff members
    - player vendors cannot be invisible
  * don't let customers buy items which have already been removed from the
    player vendor (patch from Kai Sassmannshausen)
  * no effects on hidden staff (merge from RunUO 2.0 subversion
    repository, revisions 156)
  * don't let scissors cut locked down items (merge from RunUO 2.0
    subversion repository, revisions 191)
  * fixed drop exploit (merge from RunUO 2.0 subversion repository,
    revisions 233)
  * support client version 6.0.5.0 (merge from RunUO 2.0 subversion
    repository, revisions 266)
  * use .NET 2.0 for 32 bit build in the binary distribution

 -- Max Kellermann <max@duempel.org>  Tue, 25 Dec 2007 12:04:09 +0100

sunuo (0.5.0) unstable; urgency=low

  * support client version 6 and newer (merge from RunUO 2.0 subversion
    repository, revisions 188, 193, 194, 197)
  * Makefile: add "./lib/" to MONO_PATH
  * Makefile: include "run64.bat" in distribution

 -- Max Kellermann <max@duempel.org>  Mon, 20 Aug 2007 17:58:50 +0200

sunuo (0.4.4) unstable; urgency=low

  * wake up TimerThread during shutdown
  * make Account.Limit configurable (patch by Maik Greubel)
  * new feature "quick-local-connect": do not send relay packets, re-use
    the login connection for gameplay
  * run.sh: kill SunUO process on exit
  * run.sh: don't fall back to w32 binary
  * enable the .NET server GC by default (hint from Kai Sassmannshausen)
  * upgraded Npgsql.dll to 1.0
  * ship the 64 bit / .NET 2.0 binaries for Windows
  * use xmlto instead of xsltproc + docbook-xsl

 -- Max Kellermann <max@duempel.org>  Wed, 20 Jun 2007 11:46:41 +0200

sunuo (0.4.3) unstable; urgency=low

  * make automatic world save interval configurable, patch by Maik Greubel
  * documentation updates
    - server-list configuration
  * manual locking
    - TimerChangeEntry.m_InstancePool
    - NetState.m_Disposed
  * fixed C# compiler warnings
    - use IPAddress.GetAddressBytes() instead of IPAddress.Address
    - use ScriptCompiler.Libraries instead of ScriptCompiler.Assemblies
    - don't set Listener.Port in Misc/ServerList.cs
    - ignore local port in Encryption

 -- Max Kellermann <max@duempel.org>  Tue,  9 Jan 2007 09:14:50 +0100

sunuo (0.4.2) unstable; urgency=low

  * added interface IPacket, implemented by class Packet
  * implemented abstract packet recycler code
    - recycle MobileMoving
  * continued memory optimizations
    - eliminated most DateTime.Now calls
    - all other DateTime.Now calls are converted to DateTime.UtcNow to avoid
      time zone lookups
    - reuse timer in spawner engine
    - early return when HarrowerTentacles have no drain targets
    - reuse TimerChangeEntry objects
    - NameList: trim strings in initialization, not on every request
  * manual synchronization of TimerThread.m_ChangeQueue
  * obsolete APIs
    - Core.ExePath
    - Core.BaseDirectory
    - ScriptCompiler.Assemblies
    - CrashedEventArgs.Close
  * moved filesystem paths from Core to Core.Config
    - LogDirectory
    - CacheDirectory
  * removed Core.LocalDirectoryInfo, Core.BaseDirectoryInfo
  * login: removed EventSink.Crashed

 -- Max Kellermann <max@duempel.org>  Fri,  5 Jan 2007 00:53:42 +0100

sunuo (0.4.1) unstable; urgency=low

  * lock m_Accounts in Accounting.Accounts
  * fixed account database corruption bug
  * show correct client number in disconnect message

 -- Max Kellermann <max@duempel.org>  Mon, 25 Dec 2006 01:46:23 +0100

sunuo (0.4.0) unstable; urgency=low

  * implemented a main loop profiler
  * refuse to load objects with serial -1
  * transactional world save
  * non-interactive world repair
  * reserve 10% space in mobile/item hash tables
  * script debugging is always enabled
  * explicit TileData initialization
  * added Config.Network
    - configure Listeners
  * Listener.Slice() returns null when idle
  * ScriptCompiler ignores control directories of common revision control
    systems
  * bugfix in Mobile.cs: send poison message only for standard poisons
    (patch from Kai Sassmannshausen)
  * login: optionally limit the number of account creations per IP (patch
    from Patrick Reinhardt)
  * disable some of the string.Intern() calls because they were poisoning
    the intern tables
  * synchronize NetState.FinishDispose()
  * implemented feature "manual-gc" for manual garbage collection
  * parsing boolean values from the configuration file is now strict
  * enable AutoAccountCreation by default
  * added legacy scripts
    - use log4net for logging
    - removed Address from ServerList
    - default to server list from Config.GameServerListConfig

src/Config.cs
/*
 * SunUO
 * $Id$
 *
 * (c) 2005-2006 Max Kellermann <max@duempel.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; version 2 of the License.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

using System;
using System.Collections;
using System.IO;
using System.Xml;
using System.Net;

namespace Server.Config {
	class Parser {
		public static bool ParseBool(string value, bool defaultValue) {
			if (value == null || value == "")
				return defaultValue;

			if (value == "true" || value == "on" || value == "yes")
				return true;

			if (value == "false" || value == "off" || value == "no")
				return false;

			throw new FormatException("Cannot parse boolean value");
		}

		public static string GetElementString(XmlElement parent, string tag) {
			XmlNodeList nl = parent.GetElementsByTagName(tag);
			if (nl.Count == 0)
				return null;
			return nl[0].InnerText;
		}

		public static bool GetElementBool(XmlElement parent, string tag,
										  bool defaultValue) {
			if (parent == null)
				return defaultValue;
			XmlNodeList nl = parent.GetElementsByTagName(tag);
			if (nl.Count == 0)
				return defaultValue;
			string value = ((XmlElement)nl[0]).GetAttribute("value");
			return ParseBool(value, true);
		}

		public static int GetElementInt(XmlElement parent, string tag, int defaultValue)
		{
			if (parent == null)
				return defaultValue;
			XmlNodeList nl = parent.GetElementsByTagName(tag);
			if (nl.Count == 0)
				return defaultValue;
			string value = ((XmlElement)nl[0]).GetAttribute("value");
			try
			{
				return Int32.Parse(value);
			}
			catch
			{
				return defaultValue;
			}
		}
	}

	public class Features {
		private Hashtable m_Table = new Hashtable();

		public bool this[string name] {
			get {
				return m_Table.Contains(name);
			}
			set {
				if (value)
					m_Table[name] = true;
				else
					m_Table.Remove(name);
			}
		}
	}

	public class Library {
		private string name;
		private DirectoryInfo sourcePath;
		private FileInfo binaryPath;
		private bool disabled = false;
		private string[] ignoreSources;
		private string[] ignoreTypes;
		private string[] depends;
		private string[] overlays;
		private int warningLevel = -1;

		private static string[] CollectStringArray(XmlElement parent,
												   string tag, string attr) {
			ArrayList al = new ArrayList();
			foreach (XmlElement el in parent.GetElementsByTagName(tag)) {
				string n = el.GetAttribute(attr);
				if (n != null && n != "")
					al.Add(n);
			}

			return al.Count == 0
				? null
				: (string[])al.ToArray(typeof(string));
		}
		private static string[] LowerStringArray(string[] src) {
			if (src == null)
				return null;

			string[] dst = new string[src.Length];
			for (uint i = 0; i < src.Length; i++)
				dst[i] = src[i].ToLower();

			return dst;
		}

		public Library(string _name) {
			name = _name;
		}

		public Library(string _name, DirectoryInfo _path) {
			name = _name;
			sourcePath = _path;
		}

		public Library(string _name, FileInfo _path) {
			name = _name;
			binaryPath = _path;
		}

		public void Load(XmlElement libConfigEl) {
			string sourcePathString = Parser.GetElementString(libConfigEl, "path");
			if (sourcePathString != null) {
				if (sourcePathString.EndsWith(".dll")) {
					sourcePath = null;
					binaryPath = new FileInfo(sourcePathString);
				} else {
					sourcePath = new DirectoryInfo(sourcePathString);
					binaryPath = null;
				}
			}

			ignoreSources = CollectStringArray(libConfigEl, "ignore-source", "name");
			ignoreTypes = CollectStringArray(libConfigEl, "ignore-type", "name");
			overlays = LowerStringArray(CollectStringArray(libConfigEl, "overlay", "name"));
			depends = LowerStringArray(CollectStringArray(libConfigEl, "depends", "name"));

			string disabledString = libConfigEl.GetAttribute("disabled");
			disabled = Parser.ParseBool(disabledString, false);

			string warnString = libConfigEl.GetAttribute("warn");
			if (warnString != null && warnString != "")
				warningLevel = Int32.Parse(warnString);
		}

		public string Name {
			get { return name; }
		}
		public DirectoryInfo SourcePath {
			get { return sourcePath; }
		}
		public FileInfo BinaryPath {
			get { return binaryPath; }
		}
		public bool Exists {
			get {
				return (sourcePath != null && sourcePath.Exists) ||
					(binaryPath != null && binaryPath.Exists);
			}
		}
		public bool Disabled {
			get { return disabled; }
			set { disabled = value; }
		}
		public int WarningLevel {
			get { return warningLevel; }
		}
		public string[] Overlays {
			get { return overlays; }
		}
		public string[] Depends {
			get { return depends; }
		}

		public bool GetIgnoreSource(string filename) {
			if (ignoreSources == null)
				return false;

			foreach (string ign in ignoreSources)
				/* XXX: better check */
				if (filename.EndsWith(ign))
					return true;

			return false;
		}
		public bool GetIgnoreType(Type type) {
			if (ignoreTypes == null)
				return false;

			foreach (string ign in ignoreTypes)
				if (ign == type.FullName)
					return true;

			return false;
		}
	}

	public class Network {
		private IPEndPoint[] m_Bind;

		public Network() {
		}

		public Network(XmlElement networkEl) {
			XmlNodeList nl = networkEl.GetElementsByTagName("bind");
			if (nl.Count == 0) {
				DefaultBind();
			} else {
				m_Bind = new IPEndPoint[nl.Count];
				for (int i = 0; i < nl.Count; ++i)
					m_Bind[i] = ParseEndPoint((XmlElement)nl[i]);
			}
		}

		private void DefaultBind() {
			m_Bind = new IPEndPoint[1];
			m_Bind[0] = new IPEndPoint(IPAddress.Any, 2593);
		}

		private static IPEndPoint ParseEndPoint(XmlElement el) {
			IPAddress address = IPAddress.Any;
			int port = 2593;

			string addressString = el.GetAttribute("address");
			if (addressString != null && addressString != "")
				address = IPAddress.Parse(addressString);

			string portString = el.GetAttribute("port");
			if (portString != null && portString != "")
				port = Int32.Parse(portString);

			return new IPEndPoint(address, port);
		}

		public IPEndPoint[] Bind {
			get { return m_Bind; }
		}
	}

	public class Login {
		private bool ignoreAuthID, autoCreateAccounts;
		private string accountDatabase;
		private int maxCreatedAccountsPerIP, maxLoginsPerIP, maxCharactersPerAccount;
		private Hashtable superClients = new Hashtable();

		public Login() {
		}

		public Login(XmlElement el) {
			ignoreAuthID = Parser.GetElementBool(el, "ignore-auth-id", false);
			autoCreateAccounts = Parser.GetElementBool(el, "auto-create-accounts", true);
			accountDatabase = Parser.GetElementString(el, "account-database");

			maxCreatedAccountsPerIP = Parser.GetElementInt(el, "max-created-accounts-per-ip", 0);
			maxLoginsPerIP = Parser.GetElementInt(el, "max-logins-per-ip", 0);
			maxCharactersPerAccount = Parser.GetElementInt(el, "max-chars-per-account", 5);

			foreach (XmlElement priv in el.GetElementsByTagName("super-client")) {
				string ip = priv.InnerText;
				if (ip == null)
					continue;
				ip = ip.Trim();
				if (ip == "")
					continue;
				superClients[ip] = true;
			}
		}

		public bool IgnoreAuthID {
			get { return ignoreAuthID; }
		}

		public bool AutoCreateAccounts {
			get { return autoCreateAccounts; }
		}

		public string AccountDatabase {
			get { return accountDatabase; }
		}

		public int MaxCreatedAccountsPerIP {
		    get { return maxCreatedAccountsPerIP; }
		}

		public int MaxLoginsPerIP {
		    get { return maxLoginsPerIP; }
		}

		public int MaxCharactersPerAccount {
			get { return maxCharactersPerAccount; }
		}

		public bool IsSuperClient(string ip) {
			return ip != null && superClients.ContainsKey(ip);
		}
	}

	public class GameServer {
		private String name;
		private IPEndPoint address;
		private bool sendAuthID, query, optional;

		public GameServer(String _name, IPEndPoint _address,
						  bool _sendAuthID, bool _query, bool _optional) {
			name = _name;
			address = _address;
			sendAuthID = _sendAuthID;
			query = _query;
			optional = _optional;
		}

		public String Name {
			get { return name; }
		}

		public IPEndPoint Address {
			get { return address; }
		}

		public bool SendAuthID {
			get { return sendAuthID; }
		}

		public bool Query {
			get { return query; }
		}

		public bool Optional {
			get { return optional; }
		}
	}

	public class GameServerList : IEnumerable {
		private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

		private ArrayList servers = new ArrayList();

		public GameServerList() {
		}

		public GameServerList(XmlElement el) {
			foreach (XmlElement gs in el.GetElementsByTagName("game-server")) {
				string name = Parser.GetElementString(gs, "name");
				if (name == null) {
					log.Warn("Game server without name ignored");
					continue;
				}

				string addressString = Parser.GetElementString(gs, "address");
				if (addressString == null) {
					log.Warn("Game server without address ignored");
					continue;
				}

				string[] splitted = addressString.Split(new char[]{':'}, 2);
				if (splitted.Length != 2) {
					log.Warn("Game server without port ignored");
					continue;
				}

				IPAddress ip;
				try {
					IPHostEntry he = Dns.Resolve(splitted[0]);
					if (he.AddressList.Length == 0) {
						log.WarnFormat("Failed to resolve {0}", splitted[0]);
						continue;
					}
					ip = he.AddressList[he.AddressList.Length - 1];
				} catch (Exception e) {
					log.Warn(String.Format("Failed to resolve {0}", splitted[0]), e);
					continue;
				}

				short port;
				try {
					port = Int16.Parse(splitted[1]);
				} catch {
					log.Warn("Invalid game server port ignored");
					continue;
				}

				IPEndPoint address = new IPEndPoint(ip, port);

				bool sendAuthID = Parser.GetElementBool(gs, "send-auth-id", false);
				bool query = Parser.GetElementBool(gs, "query", false);
				bool optional = Parser.GetElementBool(gs, "optional", false);

				servers.Add(new GameServer(name, address, sendAuthID,
										   query, optional));
			}
		}

		public IEnumerator GetEnumerator() {
			return servers.GetEnumerator();
		}

		public int Count {
			get { return servers.Count; }
		}

		public GameServer this[int index] {
			get { return (GameServer)servers[index]; }
		}

		public GameServer this[string name] {
			get {
				foreach (GameServer gs in servers) {
					if (gs.Name == name)
						return gs;
				}
				return null;
			}
		}
	}

	public class Root {
		private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

		private string filename;
		private XmlDocument document;
		private string serverName;
		private TimeSpan m_SaveInterval = TimeSpan.FromMinutes(15);
		private Features m_Features = new Features();
		private string m_BaseDirectory, m_ConfigDirectory,
			m_SaveDirectory, m_BackupDirectory, m_LogDirectory,
			m_CacheDirectory;
		private ArrayList m_DataDirectories;
		private Hashtable libraryConfig = new Hashtable();
		private Network m_Network;
		private Login loginConfig;
		private GameServerList gameServers;

		public Root(string _baseDirectory, string _filename) {
			m_BaseDirectory = _baseDirectory;
			m_ConfigDirectory = Path.Combine(m_BaseDirectory, "Data");
			m_SaveDirectory = Path.Combine(m_BaseDirectory, "Saves");
			m_BackupDirectory = Path.Combine(m_BaseDirectory, "Backups");

			DirectoryInfo base_dir = new DirectoryInfo(m_BaseDirectory);
			DirectoryInfo var_dir = base_dir.CreateSubdirectory("var");
			m_LogDirectory = var_dir.CreateSubdirectory("log").FullName;
			m_CacheDirectory = var_dir.CreateSubdirectory("cache").FullName;

			filename = _filename;

			Defaults();
			Load();
		}

		public bool Exists {
			get { return File.Exists(filename); }
		}

		public string ServerName {
			get { return serverName; }
		}

		public TimeSpan SaveInterval {
			get { return m_SaveInterval; }
		}

		public Features Features {
			get {
				return m_Features;
			}
		}

		public string BaseDirectory {
			get {
				return m_BaseDirectory;
			}
		}

		public string ConfigDirectory {
			get { return m_ConfigDirectory; }
		}

		public string SaveDirectory {
			get { return m_SaveDirectory; }
		}

		public string BackupDirectory {
			get { return m_BackupDirectory; }
		}

		public string LogDirectory {
			get { return m_LogDirectory; }
		}

		public string CacheDirectory {
			get { return m_CacheDirectory; }
		}

		public ArrayList DataDirectories {
			get { return m_DataDirectories; }
		}

		public Library GetLibrary(string name) {
			return (Library)libraryConfig[name];
		}
		public ICollection Libraries {
			get { return libraryConfig.Values; }
		}

		public Network Network {
			get { return m_Network; }
		}

		public Login Login {
			get { return loginConfig; }
		}

		public GameServerList GameServers {
			get { return gameServers; }
		}

		public XmlElement GetConfiguration(string path) {
			XmlElement el = document.DocumentElement;

			foreach (string seg in path.Split('/')) {
				XmlElement child = (XmlElement)el.SelectSingleNode(seg);
				if (child == null) {
					child = document.CreateElement(seg);
					el.AppendChild(child);
				}
				el = child;
			}

			return el;
		}

		private void Defaults() {
			Library coreConfig = new Library("core");
			libraryConfig["core"] = coreConfig;

			DirectoryInfo local = new DirectoryInfo(BaseDirectory)
				.CreateSubdirectory("local");

			/* find source libraries in ./local/src/ */
			DirectoryInfo src = local
				.CreateSubdirectory("src");
			foreach (DirectoryInfo sub in src.GetDirectories()) {
				string libName = sub.Name.ToLower();
				if (libraryConfig.ContainsKey(libName)) {
					log.WarnFormat("duplicate library '{0}' in '{1}'",
								   libName, sub.FullName);
					continue;
				}

				libraryConfig[libName] = new Library(libName, sub);
			}

			/* find binary libraries in ./local/lib/ */
			DirectoryInfo lib = local
				.CreateSubdirectory("lib");
			foreach (FileInfo libFile in lib.GetFiles("*.dll")) {
				string fileName = libFile.Name;
				string libName = fileName.Substring(0, fileName.Length - 4).ToLower();
				if (libraryConfig.ContainsKey(libName)) {
					log.WarnFormat("duplicate library '{0}' in '{1}'",
								   libName, libFile);
					continue;
				}

				libraryConfig[libName] = new Library(libName, libFile);
			}

			/* if the 'legacy' library was not found until now, load
			   the legacy scripts from ./Scripts/ */
			if (!libraryConfig.ContainsKey("legacy")) {
				DirectoryInfo legacy = new DirectoryInfo(Path.Combine(BaseDirectory,
																	  "Scripts"));
				if (legacy.Exists) {
					Library legacyConfig = new Library("legacy", legacy);
					libraryConfig[legacyConfig.Name] = legacyConfig;
				}
			}
		}

		public static void RemoveElement(XmlElement parent, string tag) {
			if (parent == null)
				return;
			XmlNodeList nl = parent.GetElementsByTagName(tag);
			XmlNode[] children = new XmlNode[nl.Count];
			for (int i = 0; i < children.Length; i++)
				children[i] = nl.Item(i);
			foreach (XmlNode child in children)
				parent.RemoveChild(child);
		}

		public static void SetElementBool(XmlElement parent, string tag,
										  bool value) {
			XmlElement el;

			RemoveElement(parent, tag);

			el = parent.OwnerDocument.CreateElement(tag);
			el.SetAttribute("value", value ? "on" : "off");
			parent.AppendChild(el);
		}

		private void Load() {
			document = new XmlDocument();
			m_DataDirectories = new ArrayList();

			if (Exists) {
				XmlTextReader reader = new XmlTextReader(filename);
				try {
					document.Load(reader);
				} finally {
					reader.Close();
				}
			} else {
				document.AppendChild(document.CreateElement("sunuo-config"));
			}

			// section "global"
			XmlElement global = GetConfiguration("global");
			if (global != null) {
				foreach (XmlNode node in global.ChildNodes) {
					if (node.NodeType != XmlNodeType.Element)
						continue;

					XmlElement el = (XmlElement)node;

					switch (node.Name) {
					case "server-name":
						serverName = el.InnerText;
						break;

					case "multi-threading":
						m_Features[node.Name]
							= Parser.ParseBool(el.GetAttribute("value"), true);
						break;

					case "feature":
						m_Features[el.GetAttribute("name")]
							= Parser.ParseBool(el.GetAttribute("value"), true);
						break;

					case "save-interval":
						double si = Convert.ToDouble( el.GetAttribute("value") );
						if(si < 1.0) {
							log.WarnFormat("Invalid value of save-interval, setting it to default");
						}
						else {
							m_SaveInterval = TimeSpan.FromMinutes(si);
							log.InfoFormat("Setting custom world save interval to {0} minutes", si);
						}
						break;

					default:
						log.WarnFormat("Invalid element global/{0}", node.Name);
						break;
					}
				}
			}

			// section "locations"
			XmlElement locations = GetConfiguration("locations");
			foreach (XmlNode node in locations.ChildNodes) {
				XmlElement el = node as XmlElement;
				if (el != null) {
					string path = el.InnerText;
					switch (el.Name) {
					case "config-dir":
						m_ConfigDirectory = path;
						break;

					case "save-dir":
						m_SaveDirectory = path;
						break;

					case "backup-dir":
						m_BackupDirectory = path;
						break;

					case "data-path":
						if (Directory.Exists(path))
							m_DataDirectories.Add(path);
						break;

					case "log-dir":
						m_LogDirectory = path;
						break;

					case "cache-dir":
						m_CacheDirectory = path;
						break;

					default:
						log.WarnFormat("Ignoring unknown location tag in {0}: {1}",
									   filename, el.Name);
						break;
					}
				}
			}

			// section "libraries"
			XmlElement librariesEl = GetConfiguration("libraries");
			foreach (XmlElement el in librariesEl.GetElementsByTagName("library")) {
				string name = el.GetAttribute("name");
				if (name == null || name == "") {
					log.Warn("library element without name attribute");
					continue;
				}

				name = name.ToLower();

				Library libConfig = (Library)libraryConfig[name];
				if (libConfig == null)
					libraryConfig[name] = libConfig = new Library(name);

				libConfig.Load(el);
			}

			if (!libraryConfig.ContainsKey("legacy"))
				libraryConfig["legacy"] = new Library("legacy");

			// section "network"
			XmlElement networkEl = GetConfiguration("network");
			m_Network = networkEl == null
				? new Network()
				: new Network(networkEl);

			// section "login"
			XmlElement loginEl = GetConfiguration("login");
			loginConfig = loginEl == null
				? new Login()
				: new Login(loginEl);

			// section "server-list"
			XmlElement serverListEl = GetConfiguration("server-list");
			if (serverListEl != null)
				gameServers = new GameServerList(serverListEl);
		}

		public void Save() {
			string tempFilename;
			if (filename.EndsWith(".xml")) {
				tempFilename = filename.Substring(0, filename.Length - 4) + ".new";
			} else {
				tempFilename = filename + ".new";
			}

			// section "locations"
			XmlElement locations = GetConfiguration("locations");
			RemoveElement(locations, "data-path");

			Hashtable dirHash = new Hashtable();

			foreach (string path in	m_DataDirectories) {
				/* check for double path */
				if (dirHash.ContainsKey(path))
					continue;
				dirHash[path] = true;

				XmlElement el = document.CreateElement("data-path");
				el.InnerText = path;
				locations.AppendChild(el);
			}

			// section "login"
			XmlElement login = GetConfiguration("login");

			SetElementBool(login, "ignore-auth-id", loginConfig.IgnoreAuthID);

			// write to file
			XmlTextWriter writer = new XmlTextWriter(tempFilename, System.Text.Encoding.UTF8);
			writer.Formatting = Formatting.Indented;

			try {
				document.Save(writer);
				writer.Close();
				File.Delete(filename);
				File.Move(tempFilename, filename);
			} catch {
				writer.Close();
				File.Delete(tempFilename);
				throw;
			}
		}
	}
}
